core/src/main/java/org/jdbi/v3/core/array/SqlArrayArgumentFactory.java

```diff
@@ -41,27 +40,41 @@
 public class SqlArrayArgumentFactory implements ArgumentFactory {
     @Override
     public Optional<Argument> build(Type type, Object value, ConfigRegistry config) {
-        Class<?> erasedType = GenericTypes.getErasedType(type);
+        return JdbiOptionals.findFirstPresent(
+            () -> tryArray(type, value, config),
+            () -> tryCollection(type, value, config)
+        );
+    }
 
-        if (!(erasedType.isArray() || Collection.class.isAssignableFrom(erasedType))) {
-            return Optional.empty();
-        }
+    private static Optional<Argument> tryArray(Type type, Object value, ConfigRegistry config) {
+        Class<?> rawType = GenericTypes.getErasedType(type);
 
-        if (value == null) {
-            return Optional.of(new NullArgument(Types.ARRAY));
+        if (rawType.isArray()) {
+            if (value == null) {
+                return Optional.of(new NullArgument(Types.ARRAY));
+            }
+
+            return config.get(SqlArrayTypes.class)
+                .findFor(rawType.getComponentType())
+                .map(arrayType -> new SqlArrayArgument<>(arrayType, value));
         }
 
```

Comment:
I'm not sure I see the improvement in this refactoring either.  It also ends up being significantly longer.  What was wrong with the original code?

Reply:
First 2 replies on the other remark were meant for this one

Reply:
If you open the old code, not being familiar with it, it takes some serious reading to figure out. The lookup variable just obscured the process. The whole method top to bottom just goes "this is either an array or a list, we're gonna cast it to whichever it is, and call 2 different flows to finish up depending on which it is". For all intents and purposes the method was just 2 functionally distinct methods clobbered together to save on lines of text, and not really structured to express its binary nature either.

I just factored the 2 methods apart and sort of normalized them. The fact that they look very different now compared to each other is a clear sign they shouldn't have been written as one whole. They happen to share a few variables and the null part, but that's all they share.

It's all about clarifying intent, making expression match function.

Reply:
oh, and IDE support becomes pretty crap too when a variable is inferred to be "either this or that, who knows" :p

Reply:
I think the intent is a lot less clear now.  Before, it read nicely, "If the type doesn't match, exit.  If it's null, bind null.  Otherwise look up the element binder.  Then, depending on whether it's an array or collection, find the element type and bind it.  One less controversial improvement might be to move the "how do I find the element type" bit to `IterableLike` which would remove the final branch from the code.

Reply:
I'll see if I can rewrite it to more of a golden middle way. I agree it is _larger_, but I generally find surface area much less of a mental burden/complexity than number of operations per method, ambiguity or sub-methods (the schroedinger variable being "one of two" types at the same time that require different handling every step of the way), and back-and-forth branching between 2 different processes that are conjoined at the hip for sharing a few lines of elementary code.

But my vote for the refactor bit (separate commit) is a soft vote, if your objection tends toward hard I can just stop caring about it, rewind the commit, and we'll move on.